### PR TITLE
Feature/support sycl queue in array constructor functions

### PR DESCRIPTION
**Pull request by diptorupd**
- [X] Have you provided a meaningful PR description?
Supersedes #1031 , #1022

Enables the `sycl_queue` keyword in the various dpnp array constructors (`empty`, `empty_like`, `full`, `full_like`, `zeros`, `zeros_like`, `ones`, `ones_like`). The passed in dpctl.SyclQueue or the one infered during type inference of the `dpnp.ndarray` is used to allocate memory.

The use of filter string in memory allocation is removed.

Improvements to the existing unit test cases and new ones.

- [X] Have you added a test, reproducer or referred to an issue with a reproducer?
- [X] Have you tested your changes locally for CPU and GPU devices?
- [X] Have you made sure that new changes do not introduce compiler warnings?
- [ ] If this PR is a work in progress, are you filing the PR as a draft?

### CODE REVIEW

File: numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty.py [hunk 1]
```diff
@@ -67,3 +91,19 @@ def func(shape):
...
+    assert c.usm_type == dummy_tensor.usm_type
+    assert c.sycl_device == dummy_tensor.sycl_device
+
+
+def test_dpnp_empty_exceptions():
+    queue = dpctl.SyclQueue()
+
+    @dpjit
+    def func2(a, q):
+        c = dpnp.empty_like(a, sycl_queue=q, device="cpu")
```
Comment (chudur-budur) [inline numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty.py:103]: Why `device="cpu"`?

File: numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty.py [hunk 2]
```diff
@@ -38,17 +39,40 @@ def func(shape):
...
+
+    if len(c.shape) == 1:
+        assert c.shape[0] == shape
     else:
-        c.sycl_device.filter_string == dpctl.SyclDevice().filter_string
+        assert c.shape == shape
+
+    assert c.dtype == dtype
+    assert c.usm_type == usm_type
+    assert c.sycl_device.filter_string == queue.sycl_device.filter_string
```
Comment (chudur-budur) [inline numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty.py:68]: Why are we comparing on `filter_string`? Why don't we do just `assert c.sycl_device == queue.sycl_device`?

File: numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty_like.py [hunk 3]
```diff
@@ -16,23 +16,20 @@
...
     @dpjit
     def func(a):
         c = dpnp.empty_like(a, dtype=dtype, usm_type=usm_type, device=device)
         return c
 
-    if isinstance(shape, int):
-        NZ = numpy.random.rand(shape)
-    else:
-        NZ = numpy.random.rand(*shape)
+    NZ = dpnp.empty(shape)
```
Comment (chudur-budur) [inline numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty_like.py:32]: The idea was to test if we can create an "empty" array with the same shape from a "non-empty" array. Now `dpnp.random.random()` works, we can use that.

File: numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty_like.py [hunk 4]
```diff
@@ -46,13 +43,89 @@ def func(a):
...
+    shape,
+    usm_type,
+    dtype,
+):
+    @dpjit
+    def func(arr):
+        c = dpnp.empty_like(arr)
+        return c
+
+    arr = dpnp.empty(shape=shape, usm_type=usm_type, dtype=dtype)
```
Comment (chudur-budur) [inline numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty_like.py:91]: We do not need to create another array to check if the `dtype`s are the same, we can just check if the returned array comes with the `dtype` that we provided in the argument.

File: numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty.py [hunk 5]
```diff
@@ -67,3 +91,19 @@ def func(shape):
     dummy_tensor = dpctl.tensor.empty(shape)
```
Comment (chudur-budur) [inline numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty.py:91]: We do not need to create another array to check if the `dtype`s are the same, we can just check if the returned array comes with the `dtype` that we provided in the argument.

**Review by chudur-budur — COMMENTED** [review]
Let's keep things simpler in these test files. For each function, we do these:
1. call dpnp function with default argument
2. call dpnp function with queue and pytest parameterized arguments
2. call dpnp function with device and pytest parameteried arguments
3. call dpnp with both queue and device (exception check)

**Review by chudur-budur — COMMENTED** [review]
Let's keep things simpler in these test files. For each function, we do these:
1. call dpnp function with default argument
2. call dpnp function with queue and pytest parameterized arguments
2. call dpnp function with device and pytest parameteried arguments
3. call dpnp with both queue and device (exception check)

File: numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty.py (hunk 1 above)
Comment (chudur-budur) [inline numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty.py:103]: Why are we testing `dpnp.empty_like()` inside `dpnp.empty()` test?

Comment (diptorupd) [reply numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty.py:103]: It is just a placeholder. The intent is to test if an exception is raised if both `sycl_queue` and `device` are specified.

File: numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty.py (hunk 2 above)
Comment (diptorupd) [reply numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty.py:68]: We can try. The reason is that they device objects will not be the same. It depends on how the `dpctlSyclDevice.__eq__` is implemented.

File: numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty_like.py (hunk 3 above)
Comment (diptorupd) [reply numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty_like.py:32]: I do not think it matters. Let us use `dpnp.ones` if you want a non-empty array. Does `dpnp.random.random` work on all architectures?

File: numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty_like.py [hunk 6]
```diff
@@ -46,13 +43,89 @@ def func(a):
...
+        func(arr)
+
+
+@pytest.mark.xfail
+def test_dpnp_empty_like_from_freevar_queue():
+    queue = dpctl.SyclQueue()
+
+    @dpjit
+    def func():
+        c = dpnp.empty_like(10, sycl_queue=queue)
```
[the remark this answers is not in the document]
Comment (diptorupd) [reply numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty_like.py:122]: Hmm. I am not passing a dimension, rather a scalar and the output array should be a zero-dimensional array like it is in NumPy. I see that dpnp does not support such a feature yet, so might as well change the test to dpnp.empty.

However, the test is for a different purpose. I accidentally discovered a bug that we cannot use a SyclQueue that is defined as a free variable inside a dpjit function. Maybe the test can be moved to tests for `DpctlSyclQueue` type.

File: numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty_like.py [hunk 7]
```diff
@@ -62,7 +135,7 @@ def func1(a):
         return c
 
     try:
-        func1(numpy.random.rand(5, 5))
+        func1(dpnp.empty((5, 5)))
```
[the remark this answers is not in the document]
Comment (diptorupd) [reply numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty_like.py:138]: Does the dpnp random function work on all devices. Let us use ones or zeros perhaps?

File: numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty.py (hunk 1 above)
Comment (diptorupd) [reply numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty.py]: typo. Fixed

File: numba_dpex/core/types/usm_ndarray_type.py [hunk 8]
```diff
@@ -66,9 +67,22 @@ def __init__(
         self.device = self.queue.sycl_device.filter_string
 
         if not dtype:
-            dummy_tensor = dpctl.tensor.empty(
-                1, order=layout, usm_type=usm_type, sycl_queue=self.queue
-            )
+            if is_fill_value_float:
```
Comment (diptorupd) [inline numba_dpex/core/types/usm_ndarray_type.py:70]: The logic does not belong to the `UsmNdArray` type. It should be moved into the dpnp overloads

**Comment by chudur-budur** [comment]
I think it's ready for merge. @diptorupd

**Pull request merged**